### DISCUSSION TITLE
issue: Problems with getting zod to work with react-hook-form

### DISCUSSION THREAD

**Discussion by Sti2nd** [react-hook-form/react-hook-form#7128] [Q&A] [answered]
### Version Number

7.20.2

### Codesandbox/Expo snack

https://codesandbox.io/s/react-hook-form-zod-resolver-ts-example-forked-w72vp

### Steps to reproduce

1. Go to the Codesandbox
2. Fill out numbers in both fields and click on submit

### Expected behaviour

The form data should be console logged, but isn't.

### What browsers are you seeing the problem on?

Firefox, Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment, marked as answer]
native input produces string by default.

**Reply by Sti2nd** [reply]
Thank you for answering! Adding `type="number"` to the input field that lacked type attribute does not help. Does it work for you?

**Reply by bluebill1049** [reply]
doesn't matter if the type is number, it will still return a string for native input, you can convert to a number with `valueAsNumber`.  

refer to the doc: https://react-hook-form.com/api/useform/register

**Reply by Sti2nd** [reply]
Ah, that explains it! Thank you for the patience and linking the documentation 💪🏿 Great library! 🥇

**Reply by nemonemi** [reply]
I'm experiencing the same thing, however, I'm using the `control`, not `register`.
Also, I have string and date fields, no number.

Is using `register` a must?

Here's a reproduction (thanks @Sti2nd)
https://codesandbox.io/s/react-hook-form-zod-resolver-ts-example-forked-somn3?file=/src/App.tsx

With the resolver it doesn't work, without it it does.

**Reply by Sti2nd** [reply]
Doesn't seem like you linked an example with Control @nemonemi ? 

As far as I know `valueAsNumber` is only supported on register so you raise an interesting question. Maybe someone else can answer.

**Reply by bluebill1049** [reply]
for controlled component, you can set local by yourself to any data type.

https://react-hook-form.com/api/usecontroller

**Reply by nemonemi** [reply]
I hear you, but how? Sorry, but which part of the controller documentation is pointing to it?

**Reply by foxart78** [reply]
I've same issue! someone have found a solution with `<Controller>`?
I have validation with zod

**Reply by foxart78** [reply]
> I've same issue! someone have found a solution with `<Controller>`? I have validation with zod

sorry: I just find the solution in documentation (https://react-hook-form.com/api/usecontroller/controller):

          <Controller
            name={'rate'}
            control={control}
            defaultValue=""
            render={({ field }) => (
              <TextField
                {...field}
                type="number"
                onChange={(e) => field.onChange(parseInt(e.target.value))} // <<<<< convert string to number
              />
            )}
          />
